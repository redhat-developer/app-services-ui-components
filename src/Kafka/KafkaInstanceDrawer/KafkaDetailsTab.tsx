import {
  Skeleton,
  TextContent,
  TextList,
  TextListItem,
  TextListItemVariants,
  TextListVariants,
} from "@patternfly/react-core";
import type { ReactChild, VoidFunctionComponent } from "react";
import { useTranslation } from "react-i18next";
import { FormatDate } from "../../shared";
import type { MarketplaceSubscription } from "../types";
import type { InstanceType } from "../utils";
import { DetailsTabAlert } from "./components/DetailsTabAlert";

/**
 * All fields marked as accepting undefined will show a skeleton loader until the value is available.
 * This is to support fetching the data from different APIs.
 */
export type KafkaDetailsTabProps = {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  expiryDate: Date | undefined;
  owner: string;
  region: string;
  instanceType: InstanceType;
  size: string | undefined;
  ingress: number | undefined;
  egress: number | undefined;
  storage: number | undefined;
  maxPartitions: number | undefined;
  connections: number | undefined;
  connectionRate: number | undefined;
  messageSize: number | undefined;
  billing: "prepaid" | MarketplaceSubscription | undefined;
  kafkaVersion: string;
};

export const KafkaDetailsTab: VoidFunctionComponent<KafkaDetailsTabProps> = ({
  id,
  createdAt,
  updatedAt,
  owner,
  region,
  expiryDate,
  instanceType,
  size,
  ingress,
  egress,
  storage,
  maxPartitions,
  connections,
  connectionRate,
  messageSize,
  billing,
  kafkaVersion,
}) => {
  const { t } = useTranslation("kafka");

  const renderTextListItem = (title: string, value: ReactChild | undefined) => (
    <>
      <TextListItem component={TextListItemVariants.dt}>{title}</TextListItem>
      {value ? (
        <TextListItem component={TextListItemVariants.dd}>{value}</TextListItem>
      ) : (
        <Skeleton
          width="50%"
          screenreaderText={t("common:skeleton_loader_screenreader_text", {
            contentName: title,
          })}
        />
      )}
    </>
  );
  return (
    <div className="mas--details__drawer--tab-content">
      {instanceType !== "standard" && expiryDate && (
        <DetailsTabAlert expiryDate={expiryDate} />
      )}
      <TextContent>
        <TextList component={TextListVariants.dl}>
          {instanceType === "standard" &&
            renderTextListItem(
              t("common:size"),
              size
                ? t("create-kafka-instance:streaming_size_value", {
                    value: size,
                  })
                : undefined
            )}
          {renderTextListItem(
            t("ingress"),
            ingress
              ? t("create-kafka-instance:ingress_value", {
                  value: ingress,
                })
              : undefined
          )}
          {renderTextListItem(
            t("egress"),
            egress
              ? t("create-kafka-instance:egress_value", {
                  value: egress,
                })
              : undefined
          )}
          {renderTextListItem(
            t("storage"),
            storage
              ? t("create-kafka-instance:storage_value", {
                  value: storage,
                })
              : undefined
          )}
          {renderTextListItem(
            t("partitions"),
            maxPartitions
              ? t("create-kafka-instance:partitions_value", {
                  value: maxPartitions,
                })
              : undefined
          )}
          {renderTextListItem(
            t("client_connections"),
            connections
              ? t("create-kafka-instance:client_connections_value", {
                  value: connections,
                })
              : undefined
          )}
          {renderTextListItem(
            t("connection_rate"),
            connectionRate
              ? t("create-kafka-instance:connection_rate_value", {
                  value: connectionRate,
                })
              : undefined
          )}
          {renderTextListItem(
            t("message_size"),
            messageSize
              ? t("create-kafka-instance:message_size_value", {
                  value: messageSize,
                })
              : undefined
          )}

          {renderTextListItem(t("common:id"), id)}
          {renderTextListItem(t("common:kafka_version"), kafkaVersion)}
          {renderTextListItem(t("common:owner"), owner)}
          {renderTextListItem(
            t("common:time_created"),
            <FormatDate date={createdAt} format={"long"} />
          )}
          {renderTextListItem(
            t("common:time_updated"),
            <FormatDate date={updatedAt} format={"long"} />
          )}
          {renderTextListItem(
            t("common:cloud_provider"),
            t("common:cloudProviders.aws")
          )}
          {renderTextListItem(t("common:region"), region)}
          {renderTextListItem(
            t("create-kafka-instance:billing.field_label"),
            (() => {
              if (billing === undefined) {
                return undefined;
              } else if (billing === "prepaid") {
                return t("create-kafka-instance:billing.prepaid_option");
              } else {
                switch (billing.marketplace) {
                  case "aws":
                    return (
                      <>
                        {t("create-kafka-instance:billing.marketplace_aws")}
                        <br />
                        {billing.subscription}
                      </>
                    );
                  case "azure":
                    return (
                      <>
                        {t("create-kafka-instance:billing.marketplace_azure")}
                        <br />
                        {billing.subscription}
                      </>
                    );
                  case "gcp":
                    return (
                      <>
                        {t("create-kafka-instance:billing.marketplace_gcp")}
                        <br />
                        {billing.subscription}
                      </>
                    );
                  case "rhm":
                    return (
                      <>
                        {t("create-kafka-instance:billing.marketplace_rh")}
                        <br />
                        {billing.subscription}
                      </>
                    );
                }
              }
            })()
          )}
        </TextList>
      </TextContent>
    </div>
  );
};
