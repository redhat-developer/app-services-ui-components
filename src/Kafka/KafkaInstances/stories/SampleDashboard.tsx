export default () => <div>Dashboard</div>;
