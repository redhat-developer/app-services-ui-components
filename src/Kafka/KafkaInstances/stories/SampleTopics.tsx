export default () => <div>Topics</div>;
