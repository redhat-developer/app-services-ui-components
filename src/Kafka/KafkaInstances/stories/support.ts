import { addHours } from "date-fns";
import { useEffect, useState } from "react";
import type { KafkaInstance } from "../../types";

const now = new Date().toISOString();
const future = addHours(new Date(), 19).toISOString();
const INSTANCES: KafkaInstance[] = [
  {
    id: "1",
    name: "foo",
    createdAt: now,
    updatedAt: now,
    expiryDate: undefined,
    owner: "owner",
    provider: "aws",
    region: "region",
    status: "provisioning",
    plan: "standard",
    size: "1",
    ingress: undefined,
    egress: undefined,
    storage: undefined,
    maxPartitions: undefined,
    connections: undefined,
    connectionRate: undefined,
    messageSize: undefined,
    billing: undefined,
  },
  {
    id: "2",
    name: "bar",
    createdAt: now,
    updatedAt: now,
    expiryDate: future,
    owner: "owner",
    provider: "aws",
    region: "region",
    status: "deleting",
    plan: "developer",
    size: "1",
    ingress: undefined,
    egress: undefined,
    storage: undefined,
    maxPartitions: undefined,
    connections: undefined,
    connectionRate: undefined,
    messageSize: undefined,
    billing: undefined,
  },
  {
    id: "3",
    name: "baz",
    createdAt: now,
    updatedAt: now,
    expiryDate: future,
    owner: "owner",
    provider: "azure",
    region: "region",
    status: "ready",
    plan: "standard",
    size: "2",
    ingress: undefined,
    egress: undefined,
    storage: undefined,
    maxPartitions: undefined,
    connections: undefined,
    connectionRate: undefined,
    messageSize: undefined,
    billing: undefined,
  },
];
export const useInstances = () => {
  const [instances, setInstances] = useState<KafkaInstance[]>();

  useEffect(() => {
    const timer = setTimeout(() => {
      setInstances(INSTANCES);
    }, 500);
    return () => clearTimeout(timer);
  }, []);
  return {
    instances,
    loading: instances === undefined,
  };
};
export const useInstance = (id: string) => {
  const { instances, loading } = useInstances();
  const instance = instances?.find((i) => i.id === id);

  return {
    instance,
    loading: loading || instance === undefined,
  };
};
