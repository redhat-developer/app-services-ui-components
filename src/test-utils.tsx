import type { RenderOptions, RenderResult } from "@testing-library/react";
import { act, render, waitFor } from "@testing-library/react";
import type { FunctionComponent, ReactElement } from "react";
import { Suspense } from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { I18nProvider } from "./I18n";
import { AlertProvider } from "../src/shared/Alerts";

const suspenseTestId = "i18n-suspense";

const AllTheProviders: FunctionComponent = ({ children }) => {
  return (
    <Router>
      <AlertProvider>
        <I18nProvider
          lng={"en"}
          resources={{
            en: {
              common: () => import("../locales/en/common.json"),
              "create-kafka-instance": () =>
                import("../locales/en/create-kafka-instance.json"),
              kafka: () => import("../locales/en/kafka.json"),
              metrics: () => import("../locales/en/metrics.json"),
              overview: () => import("../locales/en/overview.json"),
              datascienceoverview: () =>
                import("../locales/en/datascienceoverview.json"),
              kafkaoverview: () => import("../locales/en/kafkaoverview.json"),
              topic: () => import("../locales/en/topic.json"),
              apimgmtoverview: () =>
                import("../locales/en/apimgmtoverview.json"),
              "manage-kafka-permissions": () =>
                import("../locales/en/manage-kafka-permissions.json"),
              "create-topic": () => import("../locales/en/create-topic.json"),
              "message-browser": () =>
                import("../locales/en/message-browser.json"),
              "overview-v2": () => import("../locales/en/overview-v2.json"),
              "overview-v3": () => import("../locales/en/overview-v3.json"),
              "overview-v4": () => import("../locales/en/overview-v4.json"),
              "kafkaoverview-v2": () =>
                import("../locales/en/kafkaoverview-v2.json"),
              "kafkaoverview-v3": () =>
                import("../locales/en/kafkaoverview-v3.json"),
              "connection-tab": () =>
                import("../locales/en/connection-tab.json"),
              "service-registry": () =>
                import("../locales/en/service-registry.json"),
              "connection-tab-p1": () =>
                import("../locales/en/connection-tab-p1.json"),
            },
            it: {
              common: () => Promise.resolve({ delete: "Elimina" }),
            },
          }}
          debug={false}
        >
          <Suspense fallback={<div data-testid={suspenseTestId}>loading</div>}>
            {children}
          </Suspense>
        </I18nProvider>
      </AlertProvider>
    </Router>
  );
};

const AllTheProvidersWithRoot: FunctionComponent = ({ children }) => (
  <AllTheProviders>
    <div id={"root"}>{children}</div>
  </AllTheProviders>
);
const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

const renderDialog = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProvidersWithRoot, ...options });

async function waitForI18n(r: RenderResult) {
  await waitFor(() => {
    expect(r.queryByTestId(suspenseTestId)).not.toBeInTheDocument();
  });
}

async function waitForPopper() {
  await act(async () => {
    /* let popper do its updates */
  });
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, renderDialog, waitForI18n, waitForPopper };
