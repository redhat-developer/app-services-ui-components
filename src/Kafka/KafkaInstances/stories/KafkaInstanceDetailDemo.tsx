import {
  Breadcrumb,
  BreadcrumbItem,
  Button,
  Level,
  LevelItem,
  PageSection,
  Skeleton,
  Tab,
  TabContent,
  Tabs,
  TabsComponent,
  Title,
} from "@patternfly/react-core";
import type { ComponentType } from "react";
import { lazy, Suspense, useRef } from "react";
import {
  Link,
  Redirect,
  Route,
  Switch,
  useHistory,
  useParams,
  useRouteMatch,
} from "react-router-dom";
import { Loading } from "../../../shared";
import { useKafkaInstanceDrawer } from "../KafkaInstanceDrawer";
import { useInstance } from "./support";

const LazyDashboard = lazy(
  () =>
    new Promise<{ default: ComponentType<any> }>((resolve) =>
      setTimeout(() => resolve(import("./SampleDashboard")), 1500)
    )
);

const LazyTopics = lazy(
  () =>
    new Promise<{ default: ComponentType<any> }>((resolve) =>
      setTimeout(() => resolve(import("./SampleTopics")), 1500)
    )
);

export const KafkaInstanceDetailDemo = () => {
  const { id, tab = "dashboard" } = useParams<{
    id: string;
    tab: string | undefined;
  }>();
  const history = useHistory();
  const match = useRouteMatch({ path: "/streams/:id" });
  const { instance } = useInstance(id);
  const { openDrawer } = useKafkaInstanceDrawer();
  const dashboardRef = useRef<HTMLDivElement>(null);
  const topicsRef = useRef<HTMLDivElement>(null);
  return (
    <>
      <PageSection variant={"light"} type={"breadcrumb"}>
        <Breadcrumb>
          <BreadcrumbItem
            render={() => <Link to={"/streams"}>Kafka instances</Link>}
          />
          <BreadcrumbItem isActive>
            {instance?.name || <Skeleton width={"150px"} />}
          </BreadcrumbItem>
        </Breadcrumb>
      </PageSection>
      <PageSection variant={"light"}>
        <Level>
          <LevelItem>
            <Title headingLevel={"h1"}>
              {instance?.name || <Skeleton width={"150px"} />}
            </Title>
          </LevelItem>
          <LevelItem>
            {instance && (
              <Button onClick={() => openDrawer(instance)}>Open drawer</Button>
            )}
          </LevelItem>
        </Level>
      </PageSection>
      <PageSection variant={"light"} type={"tabs"}>
        <Tabs
          activeKey={tab}
          onSelect={(_, key) => {
            history.push({
              pathname: `${match?.url || ""}/${key}`,
            });
          }}
          component={TabsComponent.nav}
          aria-label="Tabs in the toggled separate content example"
        >
          <Tab
            eventKey={"dashboard"}
            title="Dashboard"
            tabContentId="dashboard-tab"
          />
          <Tab eventKey={"topics"} title="Topics" tabContentId="topics-tab" />
        </Tabs>
      </PageSection>
      <PageSection isFilled={true}>
        <Suspense fallback={<Loading />}>
          <Switch>
            <Route path={`${match?.path || ""}/dashboard`}>
              <TabContent
                eventKey={"dashboard"}
                id="dashboard-tab"
                ref={dashboardRef}
                aria-label="TODO"
              >
                <LazyDashboard />
              </TabContent>
            </Route>
            <Route path={`${match?.path || ""}/topics`}>
              <TabContent
                eventKey={"topics"}
                id="topics-tab"
                ref={topicsRef}
                aria-label="TODO"
              >
                <LazyTopics />
              </TabContent>
            </Route>
            <Redirect
              path={match?.path}
              exact={true}
              to={`${match?.path || ""}/dashboard`}
            />
          </Switch>
        </Suspense>
      </PageSection>
    </>
  );
};

export default KafkaInstanceDetailDemo;
