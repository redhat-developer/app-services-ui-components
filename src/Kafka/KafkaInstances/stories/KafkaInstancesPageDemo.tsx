import { KafkaInstancesPage } from "../KafkaInstancesPage";
import { useInstances } from "./support";

export const KafkaInstancesPageDemo = () => {
  const { instances } = useInstances();

  return (
    <KafkaInstancesPage
      instances={instances}
      itemCount={instances?.length}
      page={1}
      getUrlForInstance={(row) => `/streams/${row.id}`}
      onChangeOwner={() => {}}
      onClickConnectionTabLink={() => {}}
      onClickSupportLink={() => {}}
      onConnection={() => {}}
      onCreate={() => {}}
      onDelete={() => {}}
      onDetails={() => {}}
      onPageChange={() => {}}
      onQuickstartGuide={() => {}}
    />
  );
};

export default KafkaInstancesPageDemo;
