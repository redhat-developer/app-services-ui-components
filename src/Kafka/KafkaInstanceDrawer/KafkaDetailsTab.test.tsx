import { composeStories } from "@storybook/testing-react";
import { render, waitForI18n } from "../../test-utils";
import * as stories from "./KafkaDetailsTab.stories";

const {
  TrialInstanceJustCreated,
  TrialInstanceNearExpiration,
  TrialInstanceRecentlyCreated,
  StandardInstanceCreated,
  MissingDataShownAsSkeletonLoaders,
} = composeStories(stories);

describe("Details Tab", () => {
  it("renders standard instance", async () => {
    const comp = render(<StandardInstanceCreated />);
    await waitForI18n(comp);
    expect(comp.baseElement).toMatchSnapshot();
  });

  it("renders trial instance just created", async () => {
    const comp = render(<TrialInstanceJustCreated />);
    await waitForI18n(comp);
    expect(await comp.findByText("Info alert:")).toBeInTheDocument();
  });

  it("renders trial instance recently created", async () => {
    const comp = render(<TrialInstanceRecentlyCreated />);
    await waitForI18n(comp);
    expect(await comp.findByText("Warning alert:")).toBeInTheDocument();
  });

  it("renders trial instance about to expire", async () => {
    const comp = render(<TrialInstanceNearExpiration />);
    await waitForI18n(comp);
    expect(await comp.findByText("Danger alert:")).toBeInTheDocument();
  });

  it("renders prepaid billing option", async () => {
    const comp = render(<StandardInstanceCreated billing={"prepaid"} />);
    await waitForI18n(comp);
    expect(comp.getByText("Red Hat prepaid"));
  });

  it("renders aws marketplace billing option", async () => {
    const comp = render(
      <StandardInstanceCreated
        billing={{ marketplace: "aws", subscription: "aws-123" }}
      />
    );
    await waitForI18n(comp);

    expect(comp.getByText("AWS Marketplace", { exact: false }));
    expect(comp.getByText("aws-123", { exact: false }));
  });

  it("renders azure marketplace billing option", async () => {
    const comp = render(
      <StandardInstanceCreated
        billing={{ marketplace: "azure", subscription: "azure-123" }}
      />
    );
    await waitForI18n(comp);
    expect(comp.getByText("Azure Marketplace", { exact: false }));
    expect(comp.getByText("azure-123", { exact: false }));
  });

  it("renders red hat marketplace billing option", async () => {
    const comp = render(
      <StandardInstanceCreated
        billing={{ marketplace: "rhm", subscription: "rhm-123" }}
      />
    );
    await waitForI18n(comp);
    expect(comp.getByText("Red Hat Marketplace", { exact: false }));
    expect(comp.getByText("rhm-123", { exact: false }));
  });

  it("renders gcp marketplace billing option", async () => {
    const comp = render(
      <StandardInstanceCreated
        billing={{ marketplace: "gcp", subscription: "gcp-123" }}
      />
    );
    await waitForI18n(comp);
    expect(comp.getByText("GCP Marketplace", { exact: false }));
    expect(comp.getByText("gcp-123", { exact: false }));
  });

  it("renders loaders for undefined data", async () => {
    const comp = render(<MissingDataShownAsSkeletonLoaders />);
    await waitForI18n(comp);

    expect(comp.getByText("Loading Size"));
    expect(comp.getByText("Loading Ingress"));
    expect(comp.getByText("Loading Egress"));
    expect(comp.getByText("Loading Storage"));
    expect(comp.getByText("Loading Partitions"));
    expect(comp.getByText("Loading Client connections"));
    expect(comp.getByText("Loading Connection rate"));
    expect(comp.getByText("Loading Message size"));
    expect(comp.getByText("Loading Billing"));
  });
});
