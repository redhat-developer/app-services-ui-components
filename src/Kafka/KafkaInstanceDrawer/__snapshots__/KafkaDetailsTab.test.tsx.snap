// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Tab renders standard instance 1`] = `
<body>
  <div>
    <div
      class="mas--details__drawer--tab-content"
    >
      <div
        class="pf-c-content"
      >
        <dl
          class=""
          data-pf-content="true"
        >
          <dt
            class=""
            data-pf-content="true"
          >
            Size
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            1 streaming unit
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Ingress
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 50 MiB/second
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Egress
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 50 MiB/second
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Storage
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 50 GiB
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Partitions
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 1500
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Client connections
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 2000
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Connection rate
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 100 connections/second
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Message size
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            up to 1 MiB
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            ID
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            c7tkgchgb2d04561srg
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Kafka version
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            3.0.1
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Owner
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            snaithan_kafka_devexp
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Time created
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            <time
              datetime="2022-07-02T00:00:00Z"
            >
              Jul 2, 2022, 12:00 AM UTC
            </time>
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Time updated
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            <time
              datetime="2022-07-02T00:00:00Z"
            >
              Jul 2, 2022, 12:00 AM UTC
            </time>
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Cloud provider
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            Amazon Web Services
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Region
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            US East, N. Virginia
          </dd>
          <dt
            class=""
            data-pf-content="true"
          >
            Billing
          </dt>
          <dd
            class=""
            data-pf-content="true"
          >
            Red Hat prepaid
          </dd>
        </dl>
      </div>
    </div>
  </div>
  <div>
    <ul
      aria-atomic="false"
      aria-live="polite"
      class="pf-c-alert-group pf-m-toast"
    />
  </div>
</body>
`;
